hazelcast/src/test/java/com/hazelcast/client/protocol/MessageTaskSecurityTest.java

```diff
@@ -0,0 +1,185 @@
+        skip(CreateRaftGroupMessageTask.class, "Initial message while creating a Client proxy for any CP object");
+        skip(SqlExecuteMessageTask.class, "Permissions for specific objects are checked based on parsed query text");
+        skip(SqlCloseMessageTask.class, "Follow up SQL message where queryId is present");
+        skip(SqlFetchMessageTask.class, "Follow up SQL message where queryId is present");
+        skip(AuthenticationMessageTask.class, "Authentication message processing");
+        skip(AuthenticationCustomCredentialsMessageTask.class, "Authentication message processing");
+        skip(ExperimentalTpcAuthenticationMessageTask.class, "Authentication message processing");
+        skip(ExperimentalAuthenticationCustomCredentialsMessageTask.class, "Authentication message processing");
+        skip(CreateProxiesMessageTask.class, "Permissions handled in beforeProcess() method");
+    }
+
+    @Test
+    public void testGetRequiredPermissions() throws Exception {
+        Reflections reflections = new Reflections("com.hazelcast");
+        Set<Class<? extends AbstractMessageTask>> subTypes = reflections.getSubTypesOf(AbstractMessageTask.class);
+        for (Class clazz : subTypes) {
+            if (!Modifier.isAbstract(clazz.getModifiers())) {
+                assertGetRequiredPermission(clazz.getName());
+            }
+        }
+    }
+
+    @Test
+    public void testCreateProxiesOverridesBeforeProcess() throws Exception {
+        assertNotNull(CreateProxiesMessageTask.class.getDeclaredMethod("beforeProcess"));
+    }
+
+    private void assertGetRequiredPermission(String clsname) throws Exception {
+        if (SKIP_CLASS_MAP.containsKey(clsname)) {
+            return;
+        }
+        boolean returnsNull = doesGetRequiredPermissionSimpleReturnNull(clsname);
+        assertFalse(clsname + " returns null in getRequiredPermission()", returnsNull);
+    }
+
+    private boolean doesGetRequiredPermissionSimpleReturnNull(String clsname) throws NotFoundException, Exception, BadBytecode {
+        if (clsname == null) {
+            fail("Class with getRequiredPermission() method implementation not found");
+        }
+        ClassPool cp = ClassPool.getDefault();
+        ClassFile cf = cp.get(clsname).getClassFile();
+        MethodInfo minfo = cf.getMethod("getRequiredPermission");
+        if (minfo == null) {
+            return doesGetRequiredPermissionSimpleReturnNull(cf.getSuperclass());
+        }
+        CodeAttribute ca = minfo.getCodeAttribute();
+        CodeIterator ci = ca.iterator();
+        String[] ops = new String[RETURN_NULL_OPS.length];
+        int i = 0;
+        while (ci.hasNext()) {
+            i++;
+            if (i > RETURN_NULL_OPS.length) {
+                return false;
+            }
+            int index = ci.next();
+            int op = ci.byteAt(index);
+            ops[i - 1] = Mnemonic.OPCODE[op];
+        }
```

Comment:
what are we doing here? Decompiling byte code ?

Reply:
Just returning the instruction name for the given instruction (`op`); Another approach could be to have it directly in the `RETURN_NULL_OPS` constant and compare int arrays instead of the String ones. The chosen approach seemed to me more self-explaining.